sdk/program/src/hash.rs

```diff
@@ -44,14 +48,22 @@ const MAX_BASE58_LEN: usize = 44;
     AbiExample,
 )]
 #[repr(transparent)]
-pub struct Hash(pub(crate) [u8; HASH_BYTES]);
+pub struct Hash(pub(crate) HashArray);
 
 #[derive(Clone, Default)]
-pub struct Hasher {
-    hasher: Sha256,
+pub struct GenericHasher<D>
+where
+    D: Digest,
+    HashArray: From<GenericArray<u8, <D as OutputSizeUser>::OutputSize>>,
```

Comment:
It took me awhile to understand what this trait bound was doing, but it's very slick! Could you add a comment about how it's statically binding the output size of `result()` to fit within a `Hash`? I will certainly forget this next time I read through.

Reply:
I added a comment and docstring.